Added example SwerveControllerCommand

ChaoticChaosTheGreat opened the pull request.

virtuald commented on SwerveControllerCommand/Subsystems/DriveSubsystem.py:
```diff
@@ -0,0 +1,140 @@
+        )
+
+    def getPose(self):
+        # Returns the currently-estimated pose of the robot.
```
Use docstrings, see the [porting guide](https://github.com/robotpy/examples/blob/main/CONTRIBUTING.md)

ChaoticChaosTheGreat commented:
Added docstrings manually, for some reason whenever I run the code I get an error saying "No module named commands2" even though I have installed the library but the code looks good.

virtuald commented on SwerveControllerCommand/robotcontainer.py:
```diff
@@ -0,0 +1,114 @@
+            Constants.AutoConstants.kMaxAccelerationMetersPerSecondSquared
+        ).setKinematics(Constants.DriveConstants.kDriveKinematics)
+
+        """
```
By convention, docstrings are only used for describing functions. Normal `#` comments are used for all other comments.

virtuald commented on SwerveControllerCommand/robot.py:
```diff
@@ -0,0 +1,66 @@
+import commands2
+from robotcontainer import RobotContainer
+
+class Robot(TimedRobot):
```
RobotPy command robots should inherit from TimedCommandRobot

ChaoticChaosTheGreat replied:
Sorry, about that, the original file inherited from TimedRobot.

virtuald commented on SwerveControllerCommand/robot.py:
```diff
@@ -0,0 +1,66 @@
+         and running subsystem periodic() methods. This must be called from the robot's periodic
+         block in order for anything in the Command-based framework to work.
+        """
+        commands2.CommandScheduler.getInstance().run()
```
This isn't needed for TimedCommandRobot

virtuald commented:
If you're getting an error regarding commands2 not being found, then it's not installed for the current python interpreter (`python -m pip list` will tell you if robotpy-commands-v2 is installed). If it shows in the pip list for the current interpreter, that would be interesting.

Suave101 commented:
You need to add the example to the tests folder in the repository and format your code with black as per the contribution guide. I created a pull request on your repository to add the example to the test folder [here](https://github.com/ChaoticChaosTheGreat/examples/pull/1). You can format your code with black with this took I like to use called [Black Playground](https://black.vercel.app/?version=stable&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ARsAnNdAD2IimZxl1N_WlkPinBFoXIfdFTaTVkGVeHShArYj9yPlDvwBA7LhGo8BvRQqDilPtgsfdKl-ha7EFp0Ma6lY_06IceKiVsJ3BpoICJM9wU1VJLD7l3qd5xTmo78LqThf9uibGWcWCD16LBOn0JK8rhhx_Gf2ClySDJtvm7zQJ1Z-Ipmv9D7I_zhjztfi2UTVsJp7917XToHBm2EoNZqyE8homtGskFIiif5EZthHQvvOj8S2gJx8_t_UpWp1ScpIsD_Xq83LX-B956I_EBIeNoGwZZPFC5zAIoMeiaC1jU-sdOHVucLJM_x-jkzMvK8Utdfvp9MMvKyTfb_BZoe0-FAc2ZVlXEpwYgJVAGdCXv3lQT4bpTXyBwDrDVrUeJDivSSwOvT8tlnuMrXoD1Sk2NZB5SHyNmZsfyAEqLALbUnhkX8hbt5U2yNQRDf1LQhuUIOii6k6H9wnDNRnBiQHUfzKfW1CLiThnuVFjlCxQhJ60u67n3EK38XxHkQdOocJXpBNO51E4-f9z2hj0EDTu_ScuqOiC9cI8qJ4grSZIOnnQLv9WPvmCzx5zib3JacesIxMVvZNQiljq_gL7udm1yeXQjENOrBWbfBEkv1P4izWeAysoJgZUhtZFwKFdoCGt2TXe3xQ-wVZFS5KoMPhGFDZGPKzpK15caQOnWobOHLKaL8eFA-qI44qZrMQ7sSLn04bYeenNR2Vxz7hvK0lJhkgKrpVfUnZrtF-e-ubeeUCThWus4jZbKlFBe2Kroz90Elij_UZBMFCcFo0CfIx5mGlrINrTJLhERszRMMDd39XsBDzpZIYV4TcG7HoMS_IF8aMAAAxI-5uTWXbUQAAY8F7QgAAP01Vc6xxGf7AgAAAAAEWVo=)

ChaoticChaosTheGreat commented:
Thank you for the pull request, I have finally got commands2 to work, and have fixed the issues outlined. Just to recap, I have lowercased the folders and files, formatted all the code with black using Black Playground, adjusted robot.py for TimedCommandRobot instead of TimedRobot, added a few docstrings and changed a few into comments.

auscompgeek commented:
This PR is still not formatted with `black`. I would highly recommend setting up black in your IDE.

The file paths still need to be fixed to lowercase.

ChaoticChaosTheGreat commented:
Sorry about the files being renamed, git was being weird. I did paste all my files into the black playground, but I have now ran it in my IDE.

virtuald commented:
I wouldn't recommend copy/pasting into the black playground... that's way too much effort.

Just configure vscode to use black, and it will install it for you. Or, `pip install black`, then run it from the command line. Either of those will be much less error prone.

ChaoticChaosTheGreat commented:
Ye did that.

auscompgeek commented on SwerveControllerCommand/constants.py:
```diff
@@ -0,0 +1,93 @@
+import math
+
+
+class Constants:
```
We usually don't have a top-level `Constants` class – the module itself is sufficient for namespacing. See https://github.com/robotpy/examples/blob/main/FrisbeeBot/constants.py for another example.

ChaoticChaosTheGreat commented:
Within drivesubsystems.py and swervemodule.py I can't import constants. How would I fix this? Most solution just say to modify sys.path but the robotpy porting guidelines says not to.

auscompgeek commented:
> Within drivesubsystems.py and swervemodule.py I can't import constants.

Why not? It should work fine.

ChaoticChaosTheGreat commented:
No idea, it just gives me a Module Not Found error. For reference I'm using miniconda as my python interpreter. The way I'm testing my code for errors as for now is just having a separate repo where I can run the workflow from this one, but its not great.

ChaoticChaosTheGreat commented:
@auscompgeek I fixed most of the errors, but I get an error saying that commands2 has no attribute named Swerve4ControllerCommand, when the docs clearly says it does. I get this error from running the workflow file from this repo.

auscompgeek commented on SwerveControllerCommand/robotcontainer.py:
```diff
@@ -0,0 +1,109 @@
+        )
+        theta_controller.enableContinuousInput(-math.pi, math.pi)
+
+        swerve_controller_command = commands2.Swerve4ControllerCommand(
+            example_trajectory,
+            self.robotDrive.getPose,
+            # Functional interface to feed supplier
+            constants.DriveConstants.kDriveKinematics,
+            # Position controllers
+            PIDController(constants.AutoConstants.kPXController, 0, 0),
+            PIDController(constants.AutoConstants.kPYController, 0, 0),
+            theta_controller,
```
This will need to be adjusted to match https://github.com/robotpy/robotpy-commands-v2/pull/45.

auscompgeek commented on SwerveControllerCommand/robotcontainer.py:
```diff
@@ -0,0 +1,109 @@
+
+
+class RobotContainer:
+    def __init__(self):
+        """
+        The robot's subsystems
+        """
```
Triple-quoted strings are reserved for docstrings. In this case this is also a docstring, which seems to incorrectly describe this method.

auscompgeek commented on SwerveControllerCommand/robotcontainer.py:
```diff
@@ -0,0 +1,116 @@
+            constants.OIConstants.kDriverControllerPort
+        )
+
+        """
+        Configure default commands
+        """
```
Same here regarding the comment difference.
```suggestion
        # Configure default commands
```

auscompgeek commented on SwerveControllerCommand/robotcontainer.py:
```diff
@@ -0,0 +1,116 @@
+
+    def getAutonomousCommand(self):
+        """
+        Create config for trajectory
```
This doesn't appear to match the doc comment in the Java example.

auscompgeek commented on SwerveControllerCommand/constants.py:
```diff
@@ -0,0 +1,95 @@
+    kFrontRightDriveEncoderReversed = False
+    kRearRightDriveEncoderReversed = True
+
+    kDrivePeriod = TimedRobot.kDefaultPeriod
```
This file seems to be missing various important comments from the upstream example.

auscompgeek reviewed (changes requested):
I notice this needs merge conflicts addressed.
